### ISSUE TITLE
tried "Android PubSub Sample" step by step but failed. always reconnecting.

### ISSUE THREAD

**Issue by MaDragon** [awslabs/aws-sdk-android-samples#92]
Hi there,

I followed the tutorial, my region is US_WEST_2.
// IoT endpoint
    // AWS Iot CLI describe-endpoint call returns: XXXXXXXXXX.iot.<region>.amazonaws.com
    private static final String CUSTOMER_SPECIFIC_ENDPOINT = "********.iot.us-west-2.amazonaws.com";
 // Cognito pool ID. For this app, pool needs to be unauthenticated pool with
    // AWS IoT permissions.
    private static final String COGNITO_POOL_ID = "us-west-2:c3a66435-4618-44b8-95f1-ae55a0a40799";
    // Name of the AWS IoT policy to attach to a newly created certificate
    private static final String AWS_IOT_POLICY_NAME = "firstPolicy";

    // Region of AWS IoT
    private static final Regions MY_REGION = Regions.US_WEST_2;

    // Region of AWS IoT
    private static final Regions MY_REGION = Regions.US_WEST_2;
    // Filename of KeyStore file on the filesystem
    private static final String KEYSTORE_NAME = "iot_keystore";
    // Password for the private key in the KeyStore
    private static final String KEYSTORE_PASSWORD = "password";
    // Certificate and key aliases in the KeyStore
    private static final String CERTIFICATE_ID = "default";

I run the app and click the connect button, it shown up "connecting" once and "reconnecting" all the time.

My aim is use android phone as an Iot device to send sensor data to AWS IOT, so I tried this app.But it failed at this stage.
What I have missed??

It is not very clearly to me why I should to do these steps to connect AWS IOT via MQTT, didnt find any well documentations 

I followed the readme.md, and I was confused about step 3 and step 4.

In step 3, I followed 3.i, 3.ii, 3.iii, 3.iv. 
in 3.iv, Cognito_PoolNameAuth_Role andCognito_PoolNameUnauth_Role had already attached default role.
I saw the policy in default Unauth_Role is not the same as the example policy in 3.v, so I copyed the example one as Unauth_Role02, attached to Unauth_Role02. is this operation correct?

In step 4, I created a policy in IoT console, and edit the policy docment as the 4.v,
![image](https://cloud.githubusercontent.com/assets/3356673/20629030/b7128136-b329-11e6-9ac5-ba9db7735e76.png).  is this operation correct?

the policy name in AWS IoT is "firstPolicy", so I changed 
private static final String AWS_IOT_POLICY_NAME = "firstPolicy";

is this correct?
And then I skiped "Creating a Keystore for Use with AWS IoT" selction and run the app, click the connect button, always "reconnecting"

Any suggestion? @mlshon @wdane @rohandubal

**Comment by hzbcdut** [comment]
Me too. I am confused.

**Comment by hzbcdut** [comment]
I followed the "README.md" step by step.   
when I first run , it always show the error:

com.amazonaws.services.iot.model.ResourceNotFoundException: Policy not found (Service: AWSIot; Status Code: 404; Error Code: ResourceNotFoundException; Request ID: 5dc91155-f320-11e6-829f-0bd17228f870)
at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:712)
at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:388)
at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:199)
at com.amazonaws.services.iot.AWSIotClient.invoke(AWSIotClient.java:3713)
at com.amazonaws.services.iot.AWSIotClient.attachPrincipalPolicy(AWSIotClient.java:432)
at com.livestar.hzb.awsiotdemo.MainActivity$1.run(MainActivity.java:212)
at java.lang.Thread.run(Thread.java:841)


below is my configuration:
       private static final String CUSTOMER_SPECIFIC_ENDPOINT = "a36tzj8xmm0m5o.iot.us-east-1.amazonaws.com";
          private static final String COGNITO_POOL_ID = "us-east-1:6ef72903-05a9-4f15-9c15-ea05b7fa80dc";
    private static final String AWS_IOT_POLICY_NAME = "AdministratorAccess";
    private static final Regions MY_REGION = Regions.US_EAST_1;
    // Filename of KeyStore file on the filesystem
    private static final String KEYSTORE_NAME = "iot_keystore";
    // Password for the private key in the KeyStore
    private static final String KEYSTORE_PASSWORD = "password";
    // Certificate and key aliases in the KeyStore
    private static final String CERTIFICATE_ID = "default";

who can help me ?

**Comment by nekot** [comment]
I run the app and click the connect button, it shown up "connecting" once and "reconnecting" all the time.

{cognito-identity, us-west-2} was not found in region metadata, trying to construct an endpoint using the standard pattern for this region: 'cognito-identity.us-west-2.amazonaws.com'.

was shown.

I prepared keystore and it seems ok when I run below. But still I'm seeing "reconnecting" issue.

keytool -list -v -keystore <keystore name>.bks -storetype bks -storepass <keystore password> -provider org.bouncycastle.jce.provider.BouncyCastleProvider -providerpath path/to/provider/jar/bcprov-jdk15on-146.jar

**Comment by shibinvadayatt** [comment]
Does someone find solution? I am still facing this issue. When I tried initially running the project with Unauthenticated policy, I had faced this issue. That time I was trying to run my device as a IoT device using MQTT protocol. Everything was fine including certificate and keysore, even though it was not connecting, it had been showing "Reconnecting" instead. Finally, I checked my region, it was 'Asia Pacific'. I modified region from ASIA PACIFIC to US-WEST-2, fortunately that solved my problem.
Now, I am facing same issue over MQTT Web socket. I have using Cognito for the authentication purpose along with Facebook SDK. My facebook SDK will help me to get my user authenticate. After successful authentication, facebook will provide a session token. When I try to connect my application through Cognito with the facebook token to AWS, I am always getting "Reconnecting" message with no error. It is very difficult for me to figure out since it does not throw any error. Please help me

**Comment by phmagic** [comment]
I'm also seeing this issue after following the README step by step.

When I press connect here's the log:
```
07-10 10:49:02.398 12308-12308/com.amazonaws.demo.androidpubsub D/com.amazonaws.demo.androidpubsub.PubSubActivity: Status = Connecting
07-10 10:49:02.798 12308-12525/com.amazonaws.demo.androidpubsub D/com.amazonaws.demo.androidpubsub.PubSubActivity: Status = Reconnecting
```

**Comment by swapnils-cuelogic** [comment]
Hi @shibinvadayatt, Did you get any solution? I am also facing the same issue.

**Comment by ravirajjak** [comment]
It  might  be  the internet connection fluctuating  or wifi.

**Comment by dlin303** [comment]
Any updates on how to troubleshoot this? I am also seeing the same issue. The sample app shows 'Connecting' once, then just gets stuck on 'Reconnecting'.

My Cognito and IoT are both in us-east-1 so it shouldn't be an issue of mismatched regions.

**Comment by J-Ahlers** [comment]
I was facing the same issue and in hindsight do find the aws-sdk-android-samples guide a bit misleading. 

**TLDR**: There are two types of 'policies' in the AWS ecosystem which are inconveniently named identical. AWS IAM policies and AWS IoT policies. The latter is whats required for the pubsub to work.

Official documentation: http://docs.aws.amazon.com/iot/latest/developerguide/create-iot-policy.html

Creating an IAM policy and linking it to the cognito_unauth role leads to a ResourceNotFoundException (because the IoT policy does not exist). In order to create one you have to add a 'thing' and then create a certificate to which you can then add an IoT policy.

**Comment by kredden** [comment]
The ReadMe file was updated with additional policy instructions.

Regards,
Kevin

**Comment by NickLink** [comment]
I beat my head up with this almost a week.

Full course of action ->
After succesfull login you will have a jwt token

    String idToken = cognitoUserSession.getIdToken().getJWTToken();

put it into a map

    Map<String, String> logins = new HashMap<String, String>(); 
    //fill it with Cognito User token
    logins.put("cognito-idp.<REGION>.amazonaws.com/<COGNITO_USER_POOL_ID>", idToken);

then use it to set in two places (not stated in any documentation!)

    CognitoCachingCredentialsProvider credentialsProvider = new 
    CognitoCachingCredentialsProvider(context, IDENTITY_POOL_ID, REGION);
    credentialsProvider.setLogins(logins);

and

    AmazonCognitoIdentity cognitoIdentity = new AmazonCognitoIdentityClient(credentialsProvider);
    GetIdRequest getIdReq = new GetIdRequest();
    getIdReq.setLogins(logins); //or if you have already set provider logins just use credentialsProvider.getLogins()
    getIdReq.setIdentityPoolId(COGNITO_POOL_ID);
    GetIdResult getIdRes = cognitoIdentity.getId(getIdReq);

after that you still nedd to make some call

    AttachPrincipalPolicyRequest attachPolicyReq = new AttachPrincipalPolicyRequest(); //in docs it called AttachPolicyRequest but it`s wrong
    attachPolicyReq.setPolicyName("allAllowed"); //name of your IOTAWS policy
    attachPolicyReq.setPrincipal(getIdRes.getIdentityId());
    new AWSIotClient(credentialsProvider).attachPrincipalPolicy(attachPolicyReq);

and only after that you can enable connect button and continue like that

    mqttManager.connect(credentialsProvider, new AWSIotMqttClientStatusCallback() {

Really for this small piece of code i spent a lot of time...

**Comment by kvasukib** [comment]
Hi @NickLink,

We are really sorry for the inconvenience caused. We will update the README of the app with the sufficient information and make updates to the sample accordingly.

**Comment by t0mmyz7** [comment]
Hi guys

I ran into the same issue here where the sampleApp would not connect.  Your README need further updating yet:

1. The IoT Policy should include "iot:connect" , i am not sure if this is a new limitation that was introduced with the new pricing model for iot or not. Anyway without this, no connection is permitted.

` 
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:Publish",
        "iot:Receive",
        "iot:subscribe",
        "iot:connect"
      ],
      "Resource": "arn:aws:iot:us-east-1:14223477777:/your/topic"
    }
  ]
}
`
2. You should specify that a thing must be created, and the certificate (with the attached create policy) should be attached to the "thing"

Now this means that your unauthenticated access solution does not work because the app creates a certificate and a policy, but it does not attach it to a thing.

Am I correct here?

**Comment by shibinvadayatt** [comment]
@swapnils-cuelogic. Yes. The issue was policy. Create an I oT policy on server and attach that right after authentication. That policy must have permissions to access IoT services. 
Follow NickLink's answer and if you are still facing the same issue. Please write to me shibinfrancis@live.com

BR/Shibin Francis

**Comment by ParasRawat** [comment]
Please provide a detailed read me file regarding Android Pub Sub , the current read me file is not up to date. IAM does not automatically create roles corresponding to our newly created cognito user pool. There is no idea as to how to create a general keystore.  Have been searching for the solution for a week but cannot find it.

**Comment by muellerfr** [comment]
Same issue as #131. Need to update the documentation, which we will track with issue #131.
